codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/HuffmanDecoder.java

```diff
@@ -69,84 +64,14 @@
      * @throws IOException if an I/O error occurs. In particular, an <code>IOException</code> may be
      * thrown if the output stream has been closed.
      */
-    public byte[] decode(byte[] buf) throws IOException {
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
-
-        /*
-         * The idea here is to consume whole bytes at a time rather than individual bits. node
-         * represents the Huffman tree, with all bit patterns denormalized as 256 children. Each
-         * child represents the last 8 bits of the huffman code. The parents of each child each
-         * represent the successive 8 bit chunks that lead up to the last most part. 8 bit bytes
-         * from buf are used to traverse these tree until a terminal node is found.
-         *
-         * current is a bit buffer. The low order bits represent how much of the huffman code has
-         * not been used to traverse the tree. Thus, the high order bits are just garbage.
-         * currentBits represents how many of the low order bits of current are actually valid.
-         * currentBits will vary between 0 and 15.
-         *
-         * symbolBits is the number of bits of the the symbol being decoded, *including* all those
-         * of the parent nodes. symbolBits tells how far down the tree we are. For example, when
-         * decoding the invalid sequence {0xff, 0xff}, currentBits will be 0, but symbolBits will be
-         * 16. This is used to know if buf ended early (before consuming a whole symbol) or if
-         * there is too much padding.
-         */
-        Node node = root;
-        int current = 0;
-        int currentBits = 0;
-        int symbolBits = 0;
-        for (int i = 0; i < buf.length; i++) {
-            int b = buf[i] & 0xFF;
-            current = (current << 8) | b;
-            currentBits += 8;
-            symbolBits += 8;
-            // While there are unconsumed bits in current, keep consuming symbols.
-            while (currentBits >= 8) {
-                int c = (current >>> (currentBits - 8)) & 0xFF;
-                node = node.children[c];
-                currentBits -= node.bits;
-                if (node.isTerminal()) {
-                    if (node.symbol == HpackUtil.HUFFMAN_EOS) {
-                        throw EOS_DECODED;
-                    }
-                    baos.write(node.symbol);
-                    node = root;
-                    // Upon consuming a whole symbol, reset the symbol bits to the number of bits
-                    // left over in the byte.
-                    symbolBits = currentBits;
-                }
-            }
+    public AsciiString decode(ByteBuf buf, int length) throws IOException {
+        try {
+            buf.forEachByte(buf.readerIndex(), length, processor);
+            buf.skipBytes(length);
+            return processor.end();
+        } finally {
+            processor.reset();
```

Comment:
instead of calling this method after the iteration ... what about calling it before iteration. This way you will only allocate / set state when it is needed.